ribbon.py

```diff
@@ -0,0 +1,545 @@
+            #reset marker watch time, then go
+            self.processor.last_marker_time = time.time()
+            self.processor.tracking = True
+            print "Starting"
+        if button['l1']:
+            self.processor.tracking = False
+            self.drive.move(0,0)
+            print "Stopping"
+
+    def constrain(self, val, min_val, max_val):
+        return min(max_val, max(min_val, val))
+
+    def exp(self, demand, exp):
+        # function takes a demand speed from -1 to 1 and converts it to a response value
+        # with an exponential function. exponential is -inf to +inf, 0 is linear
+        exp = 1/(1 + abs(exp)) if exp < 0 else exp + 1
+        return math.copysign((abs(demand)**exp), demand)
+
+    def run(self):
+        # Startup sequence
+        logger.info('Setup camera')
+        screen = pygame.display.get_surface()
+        self.camera = picamera.PiCamera()
+        self.camera.resolution = (self.image_width, self.image_height)
+        self.camera.framerate = self.frame_rate
+        self.camera.iso = 800
+        time.sleep(2)
+#        print self.camera.awb_gains
+        self.camera.awb_mode = 'off'
+        self.camera.awb_gains = (1, 2.4)
+        self.camera.shutter_speed = 15000
+        self.camera.video_denoise = False
+        self.camera.saturation = 20
+        self.drive.lights(True)
+        logger.info('Setup the stream processing thread')
+        # TODO: Remove dependency on drivetrain from StreamProcessor
+        self.processor = StreamProcessor(
+            screen=self.screen,
+            camera=self.camera,
+            drive=self.drive,
+            dict=self.dict
+        )
+        # To switch target colour" on the fly, use:
+        # self.processor.colour = "blue"
+        self.controls = self.setup_controls()
+        logger.info('Setting up image capture thread')
+        self.image_capture_thread = ImageCapture(
+            camera=self.camera,
+            processor=self.processor
+        )
+        pygame.mouse.set_visible(True)
+        try:
+            while not self.should_die:
+                time.sleep(0.01)
+                # TODO: Tidy this
+                if self.joystick.connected:
+                    self.joystick_handler(self.joystick.check_presses())
+                self.processor.menu = self.menu
+                if self.processor.finished:
+                    self.timeout = 0
+                if self.menu:
+                    screen.fill([0, 0, 0])
+                    colour = self.processor.colour
+                    colour_bounds = self.processor.colour_bounds[colour]
+                    #add the controls and give them their initial values
+                    for ctrl in self.controls:
+                        if not ctrl['ctrl'].active():
+                            ctrl['ctrl'].add(ctrl['index'], fade=False)
+                            i = ctrl['index']
+                            ctrl['ctrl'].value = colour_bounds[i % 2][int(i/2)]
```

Comment:
around this point, the four letters `c`, `t`, `r` and `l` start to dance around in front of my eyes and i get a headache